c2rust-analyze/src/context.rs

```diff
@@ -41,6 +41,70 @@ bitflags! {
         const OFFSET_SUB = 0x0020;
         /// This pointer can be freed.
         const FREE = 0x0040;
+
+        /// This pointer is non-null.
+        ///
+        /// [`NON_NULL`] is set (or not) when the pointer is created,
+        /// and it flows forward along dataflow edges.
+        ///
+        /// The permission is [`NON_NULL`] rather than `NULL` because we allow dropping permissions in any assignment.
+        /// This means removing a permission from a pointer's [`PermissionSet`]
+        /// must allow the pointer to take on more values, not restrict it to fewer values.
+        ///
+        /// The following should be set to [`NON_NULL`]:
+        /// * the results of [`Rvalue::Ref`] and [`Rvalue::AddressOf`]
+        /// * a dereferent (i.e. [`READ`]` | `[`WRITE`]), as otherwise this would be UB
+        /// * the argument or result of a known function like [`_.offset`] where it is UB to pass it an invalid (null) pointer
+        ///
+        /// The following should not be set to [`NON_NULL`]:
+        /// * [`core::ptr::null`]
+        /// * [`core::ptr::null_mut`]
+        /// * [`core::ptr::from_exposed_addr`] with the constant `0`
+        /// * `0 as * {const,mut} _`, a cast from the constant `0` to a pointer type
+        /// * anything else
+        ///
+        /// Non-zero but invalid pointers, such as those produced by:
+        /// * [`core::ptr::invalid`]
+        /// * [`core::ptr::invalid_mut`]
+        /// * [`NonNull::dangling`]
+        /// * [`core::ptr::from_exposed_addr`]
+        /// * int-to-ptr casts though `as` or [`core::mem::transmute`]
+        ///
+        /// will not be [`NON_NULL`] but also will not be either [`READ`]` | `[`WRITE`],
+        /// which allows us to distinguish them.
```

Comment:
I don't think it's true that invalid pointers will never be `READ` or `WRITE`.  Here's an example:
```Rust
/// Returns either a valid pointer, or a small integer (< 1000) cast to a pointer.
fn valid_or_small_int() -> *mut u8 { ... }

fn f() {
    let p: *mut u8 = valid_or_small_int();
    if p as usize < 1000 {
        // It's an integer converted to a pointer
        g1(p as usize);
    } else {
        // It's a real, valid pointer
        g2(*p);
    }
}
```
Here `p` is potentially invalid (e.g. it might be `3 as *mut u8`), but it is also `READ`, because the `else` case dereferences it.

In the future we might need a separate permission, `VALID`, meaning it's either a valid pointer or null.  Then `VALID | NON_NULL` means it's a real pointer (`&T`), `VALID` without `NON_NULL` means `Option<&T>`, and non-`VALID` means `Either<&T, usize>`.  But for now we reject any code that could produce a non-`VALID` pointer, and implicitly assume that `VALID` holds on all pointers we encounter (so only the first two cases can occur).

Reply:
Yeah, you're right about this.  I was mistakenly thinking about `READ` as must be readable, not may be read (may as in there's a chance, not may as it always capable).  I was thinking about a future `VALID` permission as well, though I was hoping we could get around it, but nevermind that.  But we'll leave that to when we have to handle any invalid pointers.  It should be `Either<&T, *T>`, though.  I'll change the docs back to forbidding invalid pointers for now.

Reply:
I Clarified some of these misconceptions I had in #868.

Reply:
Fixed in 287eaaf7b795545da904eafe62576d4a64c6e911.